### DISCUSSION TITLE
LookupPath, Path, and context

### DISCUSSION THREAD

**Discussion by verdverm** [cue-lang/cue#1364] [Ideas]
I'm not sure what my question here is, so I'll start with a motivating example. Maybe this is more of a feature request by the end.

I have the following function. One would expect `c` and `p` to be equal, without greater "context" (note that I'm not referring to cue.Context directly, but in the more general sense)

```go
func foo(val cue.Value) {
	c := "c"
	v := val.LookupPath(cue.ParsePath(cp))
	p := v.Path()
	fmt.Println(p)
}
```

However, if one calls this function with something other than a root value, the result is not what one would expect

```go
package main

import (
	"fmt"

	"cuelang.org/go/cue"
	"cuelang.org/go/cue/cuecontext"
)

const val = `
a: b: c: "c"
`

func main() {
	ctx := cuecontext.New()
	v := ctx.CompileString(val)
	b := v.LookupPath(cue.ParsePath("a.b"))
	foo(b)
}
```

This prints `a.b.c`.

How might one "reset the context" so that the output is always `c`?

Or another way, when I use a subvalue, how can I make it behave like a root value without rebuilding the value? Perhaps something like a copy-on-write without the write and rather a call to do so?

**Comment by eonpatapon** [comment]
I found myself also in this situation and my workaround is also to recompile the value to be a root value. So I'm interested to know if there is another way.

Anyway I think the current behavior makes sense. If you lookup some value I would expect it to keep the info from where it comes from originally.

**Comment by antong** [comment]
Are there use cases where you don't already know the subpath when you need it? In the example, the function foo() has the information already, so doesn't really need to call v.Path(). It could just print the subpath. Anyway, it would always be possible to calculate the necessary subpath by using the lenght of the paths, and then slice operations.

I don't know if there is a best way to rid a value of its path (losing information), but you could FillPath it into a new struct, or perhaps unify with top:
```
top := ctx.BuildExpr(ast.NewIdent("_"))
val = top.Unify(val) // now at root
```

Here is a slightly modified and extended version of your example that explores both approaches:
```
package main

import (
	"fmt"

	"cuelang.org/go/cue"
	"cuelang.org/go/cue/ast"
	"cuelang.org/go/cue/cuecontext"
)

// foo expects that v is at root
func foo(v cue.Value, p cue.Path) {
	x := v.LookupPath(p)
	fmt.Println(p, x, x.Path())
}

// bar doesn't require v to be at root,
// but instead does calculations on the
// paths. Then again, it already knows
// the subpath.
func bar(v cue.Value, p cue.Path) {
	x := v.LookupPath(p)
	xps := x.Path().Selectors()
	vps := v.Path().Selectors()
	// Here it is, but then again it is the
	// same as p.
	subpath := cue.MakePath(xps[len(vps):]...)
	fmt.Println(p, x, subpath)
}

func main() {
	ctx := cuecontext.New()
	v := ctx.CompileString(`
		a: b: c: d: e: "f"
		`)
	c := v.LookupPath(cue.ParsePath("a.b.c"))
	p := cue.ParsePath("d.e")
	bar(c, p) // d.e "f" d.e
	top := ctx.BuildExpr(ast.NewIdent("_"))
	foo(c, p)        // d.e "f" a.b.c.d.e
	c = top.Unify(c) // c is now at root
	foo(c, p)        // d.e "f" d.e
}
```

**Comment by verdverm** [comment]
There are cases, such as walking and looking for values with attributes.

Again, this is just a demonstration. The point is that one cannot get a "local context" for a value passed to a function without completely rebuilding or having to be aware that it could be from anywhere. I'm bringing it up not to find a workable way forward using the code base as it is today, but rather start a discussion and feature request to see if there is a more ergonomic way this could be done. Maybe something like a `Clone()` function could be a naive solution.

More generally, I think this could play into Structural Sharing, and how one might be able to achieve independent context without copying the entire value by working with the arcs and other internal structures that implement the lattice. Something like a copy-on-write or copy-on-request like capability. Ideally only the current node would need separation and that child nodes could remain untouched, unless they are modified through the "copied" value, which would then require a deeper copying, transparent to users of the API.

@mpvl ?

**Comment by myitcv** [comment]
(Converted the issue to be a discussion from which we can create more concrete feature requests etc).

> There are cases, such as walking and looking for values with attributes.

@verdverm  please can you share a more complete use case? I'm not totally clear why approaches like those suggested by @antong in https://github.com/cue-lang/cue/discussions/1364?converting=1#discussioncomment-1598357 are not sufficient here. Again, that's not to say they are the "right" answers, but looking at a concrete example of why they are insufficient etc will help to shape whether `Value.Clone()` or any other solution is the "right" approach.
